src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/AvoidConstArrays.cs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    using static MicrosoftNetCoreAnalyzersResources;
+
+    /// <summary>
+    /// CA1861: Avoid constant arrays as arguments. Replace with static readonly arrays.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class AvoidConstArraysAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1861";
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(RuleId,
+            CreateLocalizableResourceString(nameof(AvoidConstArraysTitle)),
+            CreateLocalizableResourceString(nameof(AvoidConstArraysMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(AvoidConstArraysDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                INamedTypeSymbol? readonlySpanType = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemReadOnlySpan1);
+                INamedTypeSymbol? functionType = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemFunc2);
+
+                // Analyzes an argument operation
+                context.RegisterOperationAction(context =>
+                {
+                    IArgumentOperation? argumentOperation;
+
+                    if (context.ContainingSymbol is IMethodSymbol method && method.MethodKind == MethodKind.StaticConstructor)
+                    {
+                        return;
+                    }
+
+                    if (context.Operation is IArrayCreationOperation arrayCreationOperation) // For arrays passed as arguments
+                    {
+                        argumentOperation = arrayCreationOperation.GetAncestor<IArgumentOperation>(OperationKind.Argument);
```

Comment:
Will `GetAncestor` go too far in niche cases similar to this?

```csharp
Console.WriteLine(() => { _ = new[] { "" }; });
```

Reply:
Maybe `argumentOperation = arrayCreationOperation.WalkUpConversion().Parent`?
Specific operation kinds can be skipped depending on what we find from the current tests.

Reply:
It's not always the 1st Parent after walk up conversion, gonna keep it as is